### PR TITLE
Homework Week 2 - mandrulli

### PR DESCRIPTION
**Pull request by mandrulliBF**

### CODE REVIEW

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_8.sql`: 
Hi there,

Great job on your homework submission! You've done a good job of following the prompt and your code is well-structured and easy to read. Here are some specific points of feedback:

1) **Coverage of the homework prompt:** Your code covers all aspects of the homework prompt. You've correctly used a full outer join to combine the data from the `daily_web_metrics` and `host_activity_reduced` tables, and you've handled the imputation of empty values in the array correctly.

2) **Logical flow:** Your solution logically follows the given homework prompt. The use of common table expressions (CTEs) to separate the data for 'yesterday' and 'today' makes the code easier to follow, and your use of the `COALESCE` function to handle null values is appropriate.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments to explain your thought process, which is excellent. The formatting is also clear and consistent, which makes the code easier to understand.

4) **Robustness and error handling:** Your code seems robust and should handle most edge cases well. However, it's worth noting that if the `daily_web_metrics` table doesn't contain data for a particular day, the `DATE_DIFF` function could return a negative number, which might cause issues. You might want to add some error handling to account for this possibility.

5) **Modularity and reusability:** Your code is modular and should be easy to reuse in other contexts. The use of CTEs makes it easy to isolate and understand specific parts of the code, and the main query could be easily adapted to work with different tables or columns.

Overall, you've done a great job on this assignment. Keep up the good work!

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_8.sql`: 
Hi,

Great job on your homework submission! You've done a good job of following the prompt and your code is well-structured and easy to read. Here are a few points of feedback:

1) **Coverage of the homework prompt:** Your code covers all aspects of the homework prompt. You've correctly used a full outer join to combine the data from the `host_activity_reduced` and `daily_web_metrics` tables, and you've handled the imputation of empty values in the array correctly.

2) **Logical flow:** Your solution logically follows the given homework prompt. The use of Common Table Expressions (CTEs) to separate the data extraction for 'yesterday' and 'today' makes the code easier to understand and follow.

3) **Readability:** Your code is very readable. You've used meaningful variable names and included comments to explain your thought process. The formatting is also clean and consistent, which makes the code easier to read.

4) **Robustness and error handling:** Your code seems robust, but there's room for improvement in error handling. For instance, you could add checks to ensure that the `date` column in the `daily_web_metrics` table is not null before using it in the `DATE_DIFF` function. This would prevent potential errors if the `date` column contains null values.

5) **Modularity and reusability:** Your code is modular and could be easily adapted for other similar tasks. The use of CTEs makes it easy to modify the data extraction process if needed, and the main query could be easily adapted to handle different metrics or time periods.

Overall, you've done a great job on this assignment! Keep up the good work. For future assignments, consider adding more error handling to make your code even more robust.

File: submission/query_3.sql
Comment (HjersmanJ) [inline submission/query_3.sql:1]: The comments on this file are a bit too numerous. While the purpose of your comments is good (explaining the why), some are a bit too verbose, such as line 27. That is best placed in either a README or PR description. Try to also have your comments be concise.

File: submission/query_6.sql [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+SELECT
+    -- If the host was already present in the past data, we use that one. If it's a new host never seen before, we user the value from the "new" data
+    COALESCE(yh.host, th.host) AS host,
+    CASE
+        -- If we have already an entry of host_activity_datelist from the past host, we just update the list by adding the new data
+        WHEN yh.host_activity_datelist IS NOT NULL THEN ARRAY[th.event_date] || yh.host_activity_datelist
+        -- If the host is new, we don't have any entry of host_activity_datelist from the past of that host, we just use the new data value
+        ELSE ARRAY[th.event_date]
+    END AS host_activity_datelist,
+    -- If new data are found for that entry, the date is the one coming from the 'today' table. But, if no new entry is found for that host in the new data, we add one year to the previous year to get the current one.
```
Comment (HjersmanJ) [inline submission/query_6.sql:35]: The purpose of these comments are great but try to reduce them. For example, for line 32, I'd rephrase to "No host_activity_datelist entry if host is new so use new value".

File: submission/query_3.sql [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    ud.user_id AS user_id,
+    dev.browser_type AS browser_type,
+    ud.event_date AS event_date
+  FROM today_events_cte AS ud
+  -- A join with the devices table is required to get the broswer type of each device_id.
+  -- Nevertheless, a LEFT outer join is used since we are only interesting in knowing the browser_type for those Device IDs used in the web_events table. Those device_id that are not present in the web_events table are useless.
+  LEFT OUTER JOIN bootcamp.devices AS dev 
+  ON ud.device_id = dev.device_id
+  -- By grouping by user_id, browser_type and event_date, it is covered the case in which the same browser_type coming from different device_id is considered as a single entry.
+  GROUP BY 1, 2, 3
```
Comment (HjersmanJ) [inline submission/query_3.sql:31]: Good job using the GROUP BY aliasing with Trino!

**Review by HjersmanJ — APPROVED** [review]
Your code is great! However, the numerous amount of comments make this somewhat difficult to read. I'd recommend moving thought processes to commits or PRs and more description in the README file. That said, I do appreciate you using comments to explain the why and logic of the code and not simply narrating it.
